File: tests/fixtures/vulnerable/taint-crossfn.js
// FIXTURE: flows that cross a FUNCTION BOUNDARY, JavaScript.
// Phase 3a stopped at the call; Phase 3b follows the value in and back out.
const db = require('./db');

// Forward: the taint arrives as a parameter and reaches a sink in here.
function runQuery(sqlText) {
  // EXPECT-FLOW sql-injection
  return db.query(sqlText);
}
function handlerForward(req) {
  runQuery("SELECT * FROM users WHERE id = " + req.query.id);
}

// Backward: the helper FETCHES the source itself and returns it.
function readName(req) {
  return req.query.name;
}
function handlerReturn(req) {
  const name = readName(req);
  // EXPECT-FLOW sql-injection
  db.query("SELECT * FROM users WHERE name = '" + name + "'");
}

File: tests/fixtures/vulnerable/taint-crossfn.py
# FIXTURE: flows that cross a function boundary, Python.
from flask import request


def run_query(cursor, sql_text):
    # EXPECT-FLOW sql-injection
    cursor.execute(sql_text)


def handler_forward(cursor):
    run_query(cursor, "SELECT * FROM users WHERE id = " + request.args["id"])


def read_name():
    return request.args["name"]


def handler_return(cursor):
    name = read_name()
    # EXPECT-FLOW sql-injection
    cursor.execute("SELECT * FROM users WHERE name = '" + name + "'")
